flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/CatalogManager.java

```diff
@@ -146,6 +152,10 @@ public CatalogManager build() {
 		}
 	}
 
+	public void setCatalogTableSchemaResolver(CatalogTableSchemaResolver schemaResolver) {
```

Comment:
I also want to pass it in the ctor, but `Parser` is constructed in `Planner`, while `Planner` needs `CatalogManager` to construct. we can't get `Parser` instance when creating `CatalogManager`. see `TableEnvironmentImpl#create`. 

A workaround approach is: using AtomicReference to hold the parser instance, code looks like:
```
public static TableEnvironmentImpl create(EnvironmentSettings settings) {

		// temporary solution until FLINK-15635 is fixed
		ClassLoader classLoader = Thread.currentThread().getContextClassLoader();

		TableConfig tableConfig = new TableConfig();

		ModuleManager moduleManager = new ModuleManager();

                 // create a parser reference first 
		AtomicReference<Parser> parserRef = new AtomicReference<>();
		CatalogManager catalogManager = CatalogManager.newBuilder()
			.classLoader(classLoader)
			.config(tableConfig.getConfiguration())
			.defaultCatalog(
				settings.getBuiltInCatalogName(),
				new GenericInMemoryCatalog(
					settings.getBuiltInCatalogName(),
					settings.getBuiltInDatabaseName()))
                          // set CatalogTableSchemaResolver Supplier instead of CatalogTableSchemaResolver
			.schemaResolverSupplier(() -> new CatalogTableSchemaResolver(parserRef.get(), settings.isStreamingMode()))
			.build();

		FunctionCatalog functionCatalog = new FunctionCatalog(tableConfig, catalogManager, moduleManager);

		Map<String, String> executorProperties = settings.toExecutorProperties();
		Executor executor = ComponentFactoryService.find(ExecutorFactory.class, executorProperties)
			.create(executorProperties);

		Map<String, String> plannerProperties = settings.toPlannerProperties();
		Planner planner = ComponentFactoryService.find(PlannerFactory.class, plannerProperties)
			.create(
				plannerProperties,
				executor,
				tableConfig,
				functionCatalog,
				catalogManager);
                // set parser reference
		parserRef.set(planner.getParser());

		return new TableEnvironmentImpl(
			catalogManager,
			moduleManager,
			tableConfig,
			executor,
			functionCatalog,
			planner,
			settings.isStreamingMode()
		);
	}
```

Users are very confused about how to create `StreamTableEnvironmentImpl` (I know some users use `StreamTableEnvironmentImpl`'s ctor to create a `StreamTableEnvironment` instead of use create method)

Reply:
argh, I missed that dependency. I can't think of a better solution now. Let's then go with the `setter` approach. Could you add a short comment on the setter why do we have it instead of passing it in ctor?

I think the first paragraph from your answer would work:
```
We do not pass it in the ctor, because we need a {@link Parser} that is constructed in a {@link Planner}.  At the same time {@link Planner} needs a {@link CatalogManager} to be constructed. Thus we can't get {@link Parser} instance when creating a {@link CatalogManager}. See {@link TableEnvironmentImpl#create}.
```